039-var-sources/proposal.md

```diff
@@ -0,0 +1,187 @@
+the credential manager. This configuration should include any credentials
+necessary for authenticating with the credential manager.
+
+A var source's `config` may use `((vars))` to obtain its own credentials,
+either by using a different var source or by using the system-level credential
+manager. These may also be provided statically at pipeline-configuration time.
+Either way, the config will be encrypted when stored in the database.
+
+## `((var))` syntax
+
+The `((var))` syntax was introduced a long while back and was never formally
+specified or documented. This RFC proposes a change to it so now's a good time
+to lay it all out.
+
+The full `((var))` syntax will be
+**`((VAR_SOURCE_NAME:SECRET_PATH.SECRET_FIELD))`**.
+
+### `VAR_SOURCE_NAME`
+
+The optional `VAR_SOURCE_NAME` segment specifies which named entry under
+`var_sources` to use for the credential lookup. If omitted (along with the
+`:`), the globally configured credential manager is used.
+
+A `VAR_SOURCE_NAME` must be a valid identifier per concourse/rfcs#(TODO).
+
+### `SECRET_PATH`
+
+The required `SECRET_PATH` segment specifies the secret to be fetched. This can
+either be a single word (`foo`) or a path (`foo/bar` or `/foo/bar`), depending
+on what lookup schemes are supported by the credential manager. For example,
+Vault and CredHub have path semantics whereas Kubernetes and Azure KeyVault
+only support simple names.
+
+For credential managers which support path-based lookup, a `SECRET_PATH`
+without a leading `/` may be queried relative to a predefined set of path
+prefixes. This is how the Vault credential manager currently works; `foo` will
+be queried under `/concourse/(team name)/(pipeline name)/foo`. See [Path lookup
+rules](#path-lookup-rules) for more information.
+
+### `SECRET_FIELD`
+
+The optional `SECRET_FIELD` specifies a field on the fetched secret to read. If
+omitted, the credential manager may choose to read a 'default field' from the
+fetched credential, if it exists. For example, the Vault credential manager
+will return the value of the `value` field if present. This is useful for
+simple single-value credentials.
+
+## Credential manager secret lookup rules
+
+Pipeline-level credential managers differ from globally-configured credential
+managers in one key way: they do not have to be limited to a particular path
+scheme.
+
+This means that credentials can be shared between teams, and credential manager
+specific settings such as ACLs may be utilized to securely share access to
+common credentials.
+
+Credential managers may still choose to have default path lookup schemes for
+convenience. This RFC makes no judgment call on this because the utility of
+this will vary between credential managers.
+
+
+# Open Questions
+
+* How should pipeline-level credential managers differ in behavior from
+  system-level credential managers, if at all?
+
+* Should we allow multiple var sources to be configured at the system-level?
```

Comment:
I would say no. Because:
1. If the second var source is also owned by the concourse ops team, then not need the second at all.
2. If the second var source is owned by team/pipeline owners, then they will have give the vault access token to concourse ops team in order to add the vault to concourse cluster, which brings a security concern, token might be leaked by concourse ops team.

With pipeline var source, a pipeline owner only needs to add his own vault's token to the system vault. This way, concourse ops doesn't know anything about users' vaults, which eliminate the security concern. This model will reduce OPs cost of concourse clusters.

Reply:
What I'm trying to figure out is if there is clear and consistent behavior that we can figure out in order to support multiple levels of `var_sources`, because that will be something we have to figure out once projects (#32) is here.

Having only a single system-level credential manager kind of feels like an artificial limitation if we're going to support multiple at the pipeline-level, so I'd like to have solid reasoning as to why.

I could imagine, for example, a Concourse installation which has a Vault credential manager as well as an EC2 IAM role-based STS credential manager (concourse/concourse#3023) for access to AWS resources. As this depends on the EC2 configuration of the `web` node, it makes sense for operators to configure it, not pipeline authors. For the same reasons outlined in this proposal, though, it seems like it would be bad to require operators to choose between Vault and STS as the latter is very specialized.

Reply:
Sounds like this turned out to be useful: https://github.com/concourse/concourse/pull/4777#issuecomment-559631169

I'll update the RFC soon!

Reply:
Oops nevermind this was a completely different question.

Reply:
Now that I'm updating the docs for `var_sources`, I'm finding it very difficult to consolidate the docs for cluster-wide credential managers, which are configured through env vars and flags, and pipeline-local `var_sources` which are configured through YAML.

If it's hard to document, it'll probably be hard to learn and use. I think it would be a lot easier to be able to document `var_sources` exclusively and then just say they can be configured cluster-wide or local to a pipeline. This will also force the issue of having to reason about multi-tiered `var_sources`, which I'm anticipating to be something we need to address in the future anyway with Projects.

So at this point I'm leaning towards unifying the cluster-wide and pipeline-local credential manager config as a single concept, `var_sources`, and defining the scoping semantics such that the local var source names 'shadow' the outer ones. We kinda planned for this already anyway in #4777.